rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,450 @@
+
+Merges are expected to be read using the standard `get` API.
+
+For each key, SlateDB maintains a history of operations, which is subject to compaction.
+
+Let's denote each operation as `OPi`. A key (`K`) that has experienced `n` changes looks like this logically (physically, the changes could be spread across the memtable and multiple SST files):
+
+```
+K:    OP1    OP2    OP3    ...    OPn
+
+earliest --------------------> latest
+```
+
+Currently, during a read operation, SlateDB always observes the latest known state of the database. In other words, it only needs to look up the latest operation. If it's a `Value`, it simply returns it, and if it's a `Tombstone`, it returns `None`. All previous operations can be ignored and will be removed by compaction at some point.
+
+```
+K:    OP1    OP2    OP3    ...    OPn
+                                   ^
+                                   |
+                                  GET
+```
+
+With the addition of a **Merge Operand**, it may be necessary to look backwards at previous values, up to the point where we encounter either a `Value` or a `Tombstone`. Everything beyond that point can be ignored.
+
+```
+K:    OP1    OP2    OP3    OP4    OPn
+            Value  Merge  Merge  Merge
+                                   ^
+                                   |
+                                  GET
+```
+
+
+In the above example, `get` should essentially return something like:
+
+```
+Merge(OP2, Merge(OP3, Merge(OP4, OPn)))
+```
+
+### Write Path & Compactions
+
+During compaction, we continuously reduce the history by merging entries according to the rules described above. This process performs partial merges along the way, combining merge operands with their base values when possible. We also merge entries in the memtable as they arrive to minimize unnecessary history storage.
+
+Compaction operates on a set of source SST files and writes to a destination SST file. The process reads all source SSTs sequentially and produces a single destination SST containing the merged results.
+
+The table below shows how different combinations of previous and new values are handled during compaction:
+
+| Previous  | New       | Result    |
+|-----------|-----------|-----------|
+| tombstone | value     | value     |
+| tombstone | merge     | value     |
+| tombstone | tombstone | tombstone |
+| value     | merge     | value     |
+| value     | value     | value     |
+| value     | tombstone | tombstone |
+| merge     | merge     | merge     |
+| merge     | value     | value     |
+| merge     | tombstone | tombstone |
+
+#### Snapshot Isolation
+
+*This part assumes that [Transactions RFC](https://github.com/slatedb/slatedb/pull/260) is already in place.*
+
+Compaction must preserve snapshot isolation to avoid affecting externally observable state. RocksDB provides a [proven solution to this problem](https://github.com/facebook/rocksdb/wiki/Merge-Operator-Implementation#compaction) that we can adopt.
```

Comment:
I was thinking about this a bit more, and it's likely we won't actually need to worry about this when compacting. That's because transactions can build on the checkpoint feature that we're adding now. The transaction's read snapshot can be associated with a db checkpoint (whatever the writer checkpoint was when the transaction was started), and so all reads would use the SSTs/SRs associated with the checkpoint. So the compactor can freely compact away data referenced by a snapshot without worrying about clobbering ongoing transactions. We still need to be careful about this for the memtable and wal tables though.